### PR TITLE
Flat address x86_64 port of Nuttx

### PR DESCRIPTION
**Pull request by sonicyang**
This is a flat address mode port of Nuttx to x86_64.
It consists the following features:
 * Runs in x86_64 long mode.
 * Configurable SSE/AVX support.
 * IRQs are managed by LAPIC(X2APIC) and IOAPIC.
 * Used TSC_DEADLINE or APIC timer for systick.
 * Pages are now maps the kernel at 4GB~, but changeable.

This kernel with ostest have been tested with 
 * Qemu/KVM on a Xeon 2630v4 machine.
 * Bochs with broadwell_ult emulation.

### CODE REVIEW

File: arch/x86_64/include/arch.h [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
```
Comment (patacongo) [inline arch/x86_64/include/arch.h]: @sonicyang All new files added to the repository must have have Apache 2.0 headers with no copyright or author information.  you can copy the header from sched/sched/sched_getcpu.c for example.

For those files that have my copyright on them, I do grant you permission to change the license to the Apache 2.0 license.

Sorry but those are Apache rules:  All new files must have the Apache 2.0 license.  Older files will eventually be converted to Apache 2.0 license as well, but that will require more time.

In general, the code looks good.  Have you ran tools/nxstyle against it?  No code can come into the repositories that does not meet the NuttX coding standard.  The nxstyle program will do some basic checks for conformance to the codeing standard.

If you do those two things, I will be happy to merge the code into the repository.

Comment (patacongo) [reply arch/x86_64/include/arch.h]: This is a large change and difficult to review online.  Once you complete the above changes, I will merge to a branch were I can do a more thorough review.  Assume there are additional issues, I will then merge that branch to master.

**Comment by patacongo** [comment]
@sonicyang with your permission, I could also help you to change the license headers.  The code looks good and I would liked to see this merged.  But we need to get Apache 2.0 licenses on the code before it can come into the repositories.

**Comment by sonicyang** [comment]
@patacongo 
Sorry, I am moving to another city this week.
I won't get a reliable network connection and workable environment until this weekend.
Yes, Please help me change these, otherwise we have to wait a till the weekend.
I would be glad to see it merged soon.

**Comment by patacongo** [comment]
Yes, I can make the changes in the licensing for you.

**Pull request merged**